lib/editor/components/pattern/EditShapePanel.js

```diff
@@ -75,33 +76,41 @@ export default class EditShapePanel extends Component<Props> {
 
   _cancelEdits = () => {
     const {activePattern, resetActiveGtfsEntity, togglePatternEditing} = this.props
+    if (this._hasEdits()) {
+      if (!window.confirm('You have unsaved shape edits. Are you sure you want to cancel and revert these changes?')) {
+        return
+      }
+    }
     togglePatternEditing()
     resetActiveGtfsEntity({
       component: 'trippattern',
       entity: activePattern
     })
   }
 
-  createFromStops = () => {
-    const {activePattern, editSettings, showConfirmModal, stops} = this.props
-    showConfirmModal({
-      title: `Create pattern shape from stops?`,
-      body: `Are you sure you want to overwrite this trip pattern?`,
-      onConfirm: () => {
-        const stopLocations = stops && activePattern.patternStops && activePattern.patternStops.length
-          ? activePattern.patternStops
-            .map((s, index) => {
-              const stop = stops.find(st => st.stop_id === s.stopId)
-              if (!stop) {
-                console.warn(`Could not locate stop with stop_id=${s.stopId}`)
-                return {lng: 0, lat: 0}
-              }
-              return {lng: stop.stop_lon, lat: stop.stop_lat}
-            })
-          : []
-        this.drawPatternFromStops(activePattern, stopLocations, editSettings.present.followStreets)
-      }
-    })
+  _generateShapeFromStops = () => {
+    const {activePattern, editSettings, stops} = this.props
+    const stopLocations = stops && activePattern.patternStops && activePattern.patternStops.length
+      ? activePattern.patternStops
+        .map((s, index) => {
+          const stop = stops.find(st => st.stop_id === s.stopId)
+          if (!stop) {
+            console.warn(`Could not locate stop with stop_id=${s.stopId}`)
+            return {lng: 0, lat: 0}
+          }
+          return {lng: stop.stop_lon, lat: stop.stop_lat}
+        })
+      : []
+    this.drawPatternFromStops(activePattern, stopLocations, editSettings.present.followStreets)
```

Comment:
There's an awful lot of non-ui logic in this method and drawPatternFromStops that ends up dispacthing some actions. Maybe it'd be best to move all of this logic to an action.